diff --git a/android/build.gradle b/android/build.gradle
index 1e629815..506c1641 100644
--- a/android/build.gradle
+++ b/android/build.gradle
@@ -7,7 +7,8 @@ buildscript {
     'androidxCore'       : '1.12.0',
     'androidxLifecycle'  : '2.2.0',
     'audioSwitch'        : '1.2.2',
-    'firebaseMessaging'  : '23.4.0'
+    'firebaseMessaging'  : '23.4.0',
+    'kotlin'             : '1.9.25'
   ]
   if (project == rootProject) {
     repositories {
@@ -18,11 +19,13 @@ buildscript {
     dependencies {
       classpath "com.android.tools.build:gradle:${versions.androidGradlePlugin}"
       classpath "com.google.gms:google-services:${versions.googleServices}"
+      classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:${versions.kotlin}"
     }
   }
 }
 
 apply plugin: 'com.android.library'
+apply plugin: 'kotlin-android'
 
 def safeExtGet(prop, fallback) {
   rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
@@ -75,6 +78,8 @@ dependencies {
   implementation "com.twilio:audioswitch:${versions.audioSwitch}"
   implementation 'com.google.android.material:material:1.1.0'
 
+  implementation project(':expo-modules-core')
+
   constraints {
     implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.8.0") {
       because("kotlin-stdlib-jdk7 is now a part of kotlin-stdlib")
