diff --git a/android/src/main/java/com/twiliovoicereactnative/ExpoApplicationLifecycleListener.kt b/android/src/main/java/com/twiliovoicereactnative/ExpoApplicationLifecycleListener.kt
new file mode 100644
index 00000000..23496867
--- /dev/null
+++ b/android/src/main/java/com/twiliovoicereactnative/ExpoApplicationLifecycleListener.kt
@@ -0,0 +1,18 @@
+package com.twiliovoicereactnative
+
+import android.app.Application
+import expo.modules.core.interfaces.ApplicationLifecycleListener
+
+class ExpoApplicationLifecycleListener : ApplicationLifecycleListener{
+    private var voiceApplicationProxy: VoiceApplicationProxy? = null
+
+    override fun onCreate(application: Application?) {
+        if (application != null) {
+            voiceApplicationProxy = VoiceApplicationProxy(application)
+        }
+
+        voiceApplicationProxy?.onCreate()
+
+        super.onCreate(application)
+    }
+}
